File: stack/stack.py
class Node:

  def __init__(self, value):
    self.value = value
    self.prev = None

class Stack:

  def __init__(self):
    self.head = None
    self.length = 0

  def pop(self):
    if self.length:
      head = self.head
      self.head = head.prev
      self.length -= 1
      return head.value
    

  def push(self, item):
    self.length += 1
    node = Node(item)
    if self.length == 1:
      self.head = node
      return
    head = self.head
    node.prev = head
    self.head = node

  def peek(self):
    return self.head.value if self.head else None
  



stack = Stack()

stack.push(7)
print(stack.head.value, 7)
stack.push(10)
print(stack.head.value, 10)
stack.push(15)
print(stack.head.value, 15)

print(stack.head.prev.value, 10)

print(stack.pop(),15)
print(stack.pop(),10)
print(stack.pop(),7)
print(stack.head.value, None)

File: stack/stack.ts
type NodeContainer<T> = {
  value: T,
  prev?: NodeContainer<T> 
}

class Stack<T> {
  public length: number
  public head: NodeContainer<T> | undefined
  
  constructor(){
    this.length = 0
    this.head = undefined
  }

  push(item: T):void{
    const node: NodeContainer<T> = { value: item }
    this.length++
    if ( this.length === 1 ){
      this.head = node
      return
    }
    const head = this.head
    node.prev = head
    this.head = node
  }

  pop(): T | undefined {
    const head = this.head
    this.head = head?.prev

    this.length--
    return head?.value
  }
  
  peek(): T | undefined{
    return this.head?.value
  }
}


const stack = new Stack<number>()

stack.push(7)
console.log(stack.head!.value, 7)
stack.push(10)
console.log(stack.head!.value, 10)
stack.push(15)
console.log(stack.head!.value, 15)

console.log(stack.head?.prev?.value, 10)

console.log(stack.pop(),15)
console.log(stack.pop(),10)
console.log(stack.pop(),7)
console.log(stack.head?.value, undefined)